apps/web/src/app/(private)/projects/[projectId]/versions/[commitUuid]/documents/[documentUuid]/logs/_components/DocumentLogs/DownloadLogsModal/useDownloadLogsModal.ts

```diff
@@ -0,0 +1,211 @@
+import useLatitudeAction from '$/hooks/useLatitudeAction'
+import { useNavigate } from '$/hooks/useNavigate'
+import { SelectableRowsHook } from '$/hooks/useSelectableRows'
+import { useToggleModal } from '$/hooks/useToogleModal'
+import { ROUTES } from '$/services/routes'
+import { usePreviewLogs } from '$/stores/previewLogs'
+import { DocumentLogFilterOptions } from '@latitude-data/core/browser'
+import { useToast } from '@latitude-data/web-ui/atoms/Toast'
+import {
+  useCurrentCommit,
+  useCurrentProject,
+} from '@latitude-data/web-ui/providers'
+import { useCallback, useEffect, useMemo, useState } from 'react'
+
+// Map() to maintain the order of the columns
+const DEFAULT_STATIC_COLUMNS = [
+  'output',
+  'id',
+  'commit.title',
+  'duration',
+  'costInMillicents',
+  'tokens',
+  'createdAt',
+]
+
+const getSelectedColumns = (columns?: Map<string, boolean>) => {
+  if (!columns) return []
+  return [...columns.entries()].filter(([, value]) => value).map(([key]) => key)
+}
+
+const MAX_IMMEDIATE_DOWNLOAD = 25
+
+export function useDownloadLogsModal({
+  selectableState,
+  filterOptions,
+}: {
+  selectableState: SelectableRowsHook
+  filterOptions: DocumentLogFilterOptions
+}) {
+  const { document: latitudeDocument } = useCurrentDocument()
+  const { toast } = useToast()
+  const navigate = useNavigate()
+  const { commit } = useCurrentCommit()
+  const { project } = useCurrentProject()
+  const previewModalState = useToggleModal()
+  const [selectedLogIds, setSelectedLogIds] = useState<(string | number)[]>([])
+  const { previewData, fetchPreview, isLoading } = usePreviewLogs({
+    documentLogIds: selectedLogIds,
+    staticColumnNames: DEFAULT_STATIC_COLUMNS,
+  })
+  const [isDownloading, setIsDownloading] = useState(false)
+  const [staticColumns, setStaticColumns] = useState<Map<string, boolean>>()
+  const [parameterColumns, setParameterColumns] =
+    useState<Map<string, boolean>>()
+
+  const { execute: executeAsyncDownload } = useLatitudeAction(
+    downloadLogsAsyncAction,
+    {
+      onSuccess: () => {
+        toast({
+          title: 'Download Started',
+          description:
+            'You will receive an email with the download link once the file is ready.',
+        })
+      },
+    },
+  )
+
+  const showModal = useCallback(() => {
+    previewModalState.onOpen()
+    setSelectedLogIds(selectableState.getSelectedRowIds())
+    fetchPreview()
+  }, [
+    fetchPreview,
+    setSelectedLogIds,
+    previewModalState.onOpen,
+    selectableState.getSelectedRowIds,
+  ])
+
+  const handleImmediateDownload = useCallback(async () => {
+    const ids = selectableState.getSelectedRowIds()
+    const selectedStaticColumns = getSelectedColumns(staticColumns)
+    const selectedParameterColumns = getSelectedColumns(parameterColumns)
+    const formData = new FormData()
+    formData.append('ids', JSON.stringify(ids))
+    formData.append('staticColumnNames', JSON.stringify(selectedStaticColumns))
```

Comment:
you've only added the selected columns feature to synchronous downloads but not async, which will require editing the downloadLogsAsyncAction – take into account query performance when doing this.